metric/held_metric.go

```diff
@@ -6,219 +6,125 @@ import (
 	"sync/atomic"
 	"time"
 
+	"github.com/cespare/xxhash/v2"
 	"github.com/ozontech/file.d/xtime"
 	"github.com/prometheus/client_golang/prometheus"
-	"github.com/segmentio/fasthash/fnv1a"
 )
 
-type heldLabels struct {
+type heldMetric[T prometheus.Metric] struct {
 	labels    []string
-	lastUsage *atomic.Int64 // unix timestamp
-
-	// can be replaced for testing
-	isObsoleteFunc func(int64, time.Duration) bool
+	lastUsage atomic.Int64 // unixnano timestamp
+	metric    T
 }
 
-func newHeldLabels(lvs []string) heldLabels {
+func newHeldMetric[T prometheus.Metric](lvs []string, metric T) *heldMetric[T] {
 	labels := make([]string, len(lvs))
+	// copy labels because they are unsafe strings
```

Comment:
if its unsafe strings, we need to deep copy strings

so we need to replace
`copy(labels, lvs)`
with
```go
labels = slices.Clone(labels)
for i := range labels {
  labels[i] = strings.Clone(labels[i])
}
```
?

Reply:
or why we copy the slice?